### PR TITLE
add optuna example: xgboost CV

### PR DESCRIPTION
**Pull request by sskarkhanis**
## Motivation
Add example of using xgboost CV and setting n_estimators for a final model after early stopping.

## Description of the changes
Added an example for xgboost CV on scikit-learn cancer dataset.

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@sskarkhanis Thank you for your PR.
A CI job failed at the format checking. Please apply `black` to the example file as follows:

```console
$ black . --line-length 99  --exclude="docs"
```

**Comment by sskarkhanis** [comment]
> @sskarkhanis Thank you for your PR.
> A CI job failed at the format checking. Please apply `black` to the example file as follows:
> 
> ```
> $ black . --line-length 99  --exclude="docs"
> ```

Hello
I've made the changes.

Thank you

File: examples/xgboost_cv.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        os.makedirs(file_path)
+    except Exception:
+        print("please create folder to store xgboost CV results")
+
+# Set Constants
+seed = 108
+n_folds = 3
+
+# Make the Optuna Sampler behave in a deterministically.
+sampler = TPESampler(seed=seed)
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:40]: The most of Optuna examples do not fix seed, so please remove this line.

File: examples/xgboost_cv.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    )
+
+    # (Optional)
+    # Print n_estimators in the output at each call to the objective function
+    print(
+        "-" * 10,
+        "Trial {} has optimal trees: {}".format(trial.number, str(xgb_cv_results.shape[0])),
+        "-" * 10,
+    )
+    print()
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:91]: I think this print statements make the output a bit verbose. How about removing it and showing `n_estimators` at the end of the execution (i.e., L109)?

```console
---------- Trial 8 has optimal trees: 1696 ----------

[I 2020-05-15 07:17:48,823] Finished trial#8 with value: 0.9882563333333333 with parameters: {'booster': 'gbtree', 'alpha': 0.11006970160449274, 'max_depth': 8, 'min_child_weight': 7, 'eta': 0.0016253741410996155, 'gamma': 0.019000120588029624, 'subsample': 0.6744265008652889, 'colsample_bytree': 0.6080708165519185, 'grow_policy': 'lossguide'}. Best is trial#7 with value: 0.991965.
---------- Trial 9 has optimal trees: 1408 ----------

[I 2020-05-15 07:17:51,772] Finished trial#9 with value: 0.9897633333333333 with parameters: {'booster': 'gbtree', 'alpha': 0.0027852366728488695, 'max_depth': 8, 'min_child_weight': 7, 'eta': 0.002435148116311601, 'gamma': 0.019866111168368436, 'subsample': 0.8959344463932998, 'colsample_bytree': 0.7192059963807262, 'grow_policy': 'depthwise'}. Best is trial#7 with value: 0.991965.

```

File: examples/xgboost_cv.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        param["gamma"] = trial.suggest_loguniform("gamma", 1e-3, 1.0)
+        param["subsample"] = trial.suggest_loguniform("subsample", 0.6, 1.0)
+        param["colsample_bytree"] = trial.suggest_loguniform("colsample_bytree", 0.6, 1.0)
+        param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])
+
+    if param["booster"] == "dart":
+        param["sample_type"] = trial.suggest_categorical("sample_type", ["uniform", "weighted"])
+        param["normalize_type"] = trial.suggest_categorical("normalize_type", ["tree", "forest"])
+        param["rate_drop"] = trial.suggest_loguniform("rate_drop", 1e-8, 1.0)
+        param["skip_drop"] = trial.suggest_loguniform("skip_drop", 1e-8, 1.0)
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:71]: How about using the same optimization settings as [`examples/pruning/xgboost_cv_integration.py` ](https://github.com/optuna/optuna/blob/master/examples/pruning/xgboost_cv_integration.py)?

```suggestion
    param = {
        "silent": 1,
        "objective": "binary:logistic",
        "eval_metric": "auc",
        "booster": trial.suggest_categorical("booster", ["gbtree", "gblinear", "dart"]),
        "lambda": trial.suggest_loguniform("lambda", 1e-8, 1.0),
        "alpha": trial.suggest_loguniform("alpha", 1e-8, 1.0),
    }

    if param["booster"] == "gbtree" or param["booster"] == "dart":
        param["max_depth"] = trial.suggest_int("max_depth", 1, 9)
        param["eta"] = trial.suggest_loguniform("eta", 1e-8, 1.0)
        param["gamma"] = trial.suggest_loguniform("gamma", 1e-8, 1.0)
        param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])
    if param["booster"] == "dart":
        param["sample_type"] = trial.suggest_categorical("sample_type", ["uniform", "weighted"])
        param["normalize_type"] = trial.suggest_categorical("normalize_type", ["tree", "forest"])
        param["rate_drop"] = trial.suggest_loguniform("rate_drop", 1e-8, 1.0)
        param["skip_drop"] = trial.suggest_loguniform("skip_drop", 1e-8, 1.0)
```

File: examples/xgboost_cv.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+# Create folder if it doesnt exist
+if not os.path.exists(file_path):
+    try:
+        os.makedirs(file_path)
+    except Exception:
+        print("please create folder to store xgboost CV results")
+
+# Set Constants
+seed = 108
+n_folds = 3
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:37]: Please use capital letters for constants.
```suggestion
SEED = 108
N_FOLDS = 3
```

File: examples/xgboost_cv.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    # Extract the best score
+    best_score = xgb_cv_results["test-auc-mean"].values[-1]
+    return best_score
+
+
+if __name__ == "__main__":
+    study = optuna.create_study(direction="maximize", sampler=sampler)
+    study.optimize(objective, n_trials=10)
+    print()
+    print(study.best_trial)
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:109]: Some examples show the best trial, but I think it's kind for users to show the formatted output because this example has many hyperparameters.

Example output:
```console
FrozenTrial(number=7, value=0.991965, datetime_start=datetime.datetime(2020, 5, 15, 7, 17, 45, 351583), datetime_complete=datetime.datetime(2020, 5, 15, 7, 17, 45, 657804), params={'booster': 'gbtree', 'alpha': 0.005432114697045233, 'max_depth': 6, 'min_child_weight': 4, 'eta': 0.25584322384906694, 'gamma': 0.01993575229218288, 'subsample': 0.7217599909772354, 'colsample_bytree': 0.8655936942560539, 'grow_policy': 'lossguide'}, distributions={'booster': CategoricalDistribution(choices=('gbtree',)), 'alpha': LogUniformDistribution(high=1.0, low=0.001), 'max_depth': IntUniformDistribution(high=9, low=1, step=1), 'min_child_weight': IntUniformDistribution(high=9, low=1, step=1), 'eta': LogUniformDistribution(high=1.0, low=0.001), 'gamma': LogUniformDistribution(high=1.0, low=0.001), 'subsample': LogUniformDistribution(high=1.0, low=0.6), 'colsample_bytree': LogUniformDistribution(high=1.0, low=0.6), 'grow_policy': CategoricalDistribution(choices=('depthwise', 'lossguide'))}, user_attrs={'n_estimators': 57}, system_attrs={}, intermediate_values={}, trial_id=7, state=TrialState.COMPLETE)
```

I suggest using a format similar to [examples/pruning/xgboost_cv_integration.py](https://github.com/optuna/optuna/blob/master/examples/pruning/xgboost_cv_integration.py).
```suggestion
    print("Best trial:")
    trial = study.best_trial

    print("  Value: {}".format(trial.value))

    print("  Params: ")
    for key, value in trial.params.items():
        print("    {}: {}".format(key, value))

    print("  Number of estimators: {}".format(trial.user_attrs["n_estimators"]))
```

File: examples/xgboost_cv.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    trial.set_user_attr("n_estimators", len(xgb_cv_results))
+
+    # Extract the best score
+    best_score = xgb_cv_results["test-auc-mean"].values[-1]
+    return best_score
+
+
+if __name__ == "__main__":
+    study = optuna.create_study(direction="maximize", sampler=sampler)
+    study.optimize(objective, n_trials=10)
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:107]: Please add ten more trials because `TPESampler` internally uses `RandomSampler` for the first ten trials.
```suggestion
    study.optimize(objective, n_trials=20)
```

File: examples/xgboost_cv.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import optuna
+from optuna.samplers import TPESampler
+import os
+import sklearn.datasets
+import sklearn.metrics
+import xgboost as xgb
+
+
+# Set path to save CV results
+file_path = "./tmp_results_xgbCV_optuna/"
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:26]: If this is a temporary directory, how about removing it at the end of the example?
```suggestion
# Set path to save CV results
file_path = "./tmp_results_xgbCV_optuna/"
```

File: examples/xgboost_cv.py (hunk 1 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py:40]: I've removed this line. 
Personally, I think it's a feature that should be highlighted in the examples; it only exists in the FAQ's and having it in an example just reinforces the reproducability aspect.

File: examples/xgboost_cv.py (hunk 2 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py:91]: Removed.

File: examples/xgboost_cv.py (hunk 3 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py:71]: Changed

File: examples/xgboost_cv.py (hunk 4 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py:37]: Changed

File: examples/xgboost_cv.py (hunk 5 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py:109]: Changed

File: examples/xgboost_cv.py (hunk 7 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py:26]: Removed.

File: examples/xgboost_cv.py (hunk 1 above)
Comment (toshihikoyanase) [reply examples/xgboost_cv.py]: I'm sorry for the delayed response.

>  I think it's a feature that should be highlighted in the examples;

That makes sense. The examples in this repository tend to prefer simplicity to the completeness (including reproducibility). For now, I'd like to follow the convention, but I appreciate your suggestion.

File: examples/xgboost_cv.py (hunk 3 above)
Comment (toshihikoyanase) [reply examples/xgboost_cv.py]: Thank you!

File: examples/xgboost_cv.py (hunk 7 above)
Comment (toshihikoyanase) [reply examples/xgboost_cv.py]: Oh, I'm sorry, but I didn't mean to remove the code. I proposed to delete the result directory at the end of the file smilarly to the other examples. IMO, the code to save the CV results of trials will help xgboost users.

I proposed the following changes. What do you think?

```python
import os
import shutil

CV_RESULT_DIR = "./xgboost_cv_results"

if not os.path.exists(file_path):
    os.mkdir(CV_RESULT_DIR)
...

    # Save cross-validation results.
    filepath = os.path.join(CV_RESULT_DIR, '{}.csv'.format(trial.number))
    xgb_cv_results.to_csv(filepath, index=False)

...

shutil.rmtree(CV_RESULT_DIR)
```

File: examples/xgboost_cv.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
+"""
+Optuna example that performs Cross Validation for cancer dataset
+using XGBoost.
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:3]: [nit] Please add a blank line after the first sentence like other examples.
```suggestion
Optuna example that performs cross-validation for cancer dataset using XGBoost.

```

File: examples/xgboost_cv.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import os
+import sklearn.datasets
+import sklearn.metrics
+import xgboost as xgb
+
+# Set Constants
+SEED = 108
+N_FOLDS = 3
+
+# Define Optuna Objective
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:27]: I think we can remove this comment since `def objective` explains this comment.
```suggestion
```

File: examples/xgboost_cv.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        dtrain=dtrain,
+        num_boost_round=10000,
+        nfold=N_FOLDS,
+        stratified=True,
+        early_stopping_rounds=100,
+        seed=SEED,
+        verbose_eval=False,
+    )
+
+    # Set n_estimators as a trial attribute; Accessible via study.trials_dataframe()
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:66]: [nit] Please add a period at the end of the sentence.
```suggestion
    # Set n_estimators as a trial attribute; Accessible via study.trials_dataframe().
```

File: examples/xgboost_cv.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    trial.set_user_attr("n_estimators", len(xgb_cv_results))
+
+    # Extract the best score
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:69]: ditto.
```suggestion
    # Extract the best score.
```

**Comment by toshihikoyanase** [comment]
@sskarkhanis I found that this PR is labeled as stale.
Please let me know if you have any troubles.

**Comment by sskarkhanis** [comment]
Hello

I had a couple of very busy weeks. I will make the changes requested in a week or so.

Thanks!

**Comment by toshihikoyanase** [comment]
@sskarkhanis Thank you for your reply! I'm looking forward to your update.

File: examples/xgboost_cv.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+    # Save cross-validation results.
+    filepath = os.path.join(CV_RESULT_DIR, '{}.csv'.format(trial.number))
+    xgb_cv_results.to_csv(filepath, index=False)
+
+    # Extract the best score.
+    best_score = xgb_cv_results["test-auc-mean"].values[-1]
+    return best_score
+
+shutil.rmtree(CV_RESULT_DIR)
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:87]: Currently, `CV_RESULT_DIR` will remove before the execution of `study.optimize`. To remove the it at the end of the execution, please move this line to the line after L101.

File: examples/xgboost_cv.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        early_stopping_rounds=100,
+        seed=SEED,
+        verbose_eval=False,
+    )
+
+    # Set n_estimators as a trial attribute; Accessible via study.trials_dataframe().
+    trial.set_user_attr("n_estimators", len(xgb_cv_results))
+
+    # Save cross-validation results.
+    filepath = os.path.join(CV_RESULT_DIR, '{}.csv'.format(trial.number))
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:80]: Please use double quates to pass the code formatter.
```suggestion
    filepath = os.path.join(CV_RESULT_DIR, "{}.csv".format(trial.number))
```

Comment (sskarkhanis) [reply examples/xgboost_cv.py]: Completed.

File: examples/xgboost_cv.py (hunk 12 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py]: Completed.

File: examples/xgboost_simple.py [hunk 14]
```diff
@@ -24,7 +24,6 @@
 import sklearn.metrics
 from sklearn.model_selection import train_test_split
 import xgboost as xgb
-
 import optuna
```
Comment (toshihikoyanase) [inline examples/xgboost_simple.py:27]: This blank line is to separate the third-party library and our library (i.e., optuna), so please revert the change.

**Comment by toshihikoyanase** [comment]
@sskarkhanis Thank you for your update. But, unfortunately, the CI jobs failed due to the out-dated `setup.py` (c.f., [checks](https://circleci.com/gh/optuna/optuna/51120?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) and [test-python35](https://circleci.com/gh/optuna/optuna/51124?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)).
This issue will be fixed if you merge the master branch.

Just FYI, but the command will be as follows:
```console
$ git merge --no-ff master
```

Or, you can rebase the branch if you want.

**Comment by sskarkhanis** [comment]
I did merge as per, 

git merge --no-ff master

but the result was the same.

**Comment by toshihikoyanase** [comment]
@sskarkhanis Hmm, this may be because the [`master` branch](https://github.com/sskarkhanis/optuna) in your forked repository is not up-to-date. (Sorry if you fetch the change locally.)

image

Could you set the `upstream` and fetch the changes in `optuna/optuna`'s `master` branch?
```console
# Set upstream
$ git checkout master
$ git remote add upstream git://github.com/optuna/optuna.git
$ git fetch upstream
$ git merge upstream/master
```
Then, please try to merge the `master` branch again.

```console
$ git merge --no-ff master
```

File: examples/xgboost_cv.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+import optuna
+
+
+# Set constants
+SEED = 108
+N_FOLDS = 3
+CV_RESULT_DIR = "./xgboost_cv_results"
+
+if not os.path.exists(file_path):
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:36]: `file_path` is not defined in this scope. I think it should be `CV_RESULT_DIR`.
```suggestion
if not os.path.exists(CV_RESULT_DIR):
```

File: examples/xgboost_cv.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    os.mkdir(CV_RESULT_DIR)
+
+# FYI: Objective functions can take additional arguments
+# (https://optuna.readthedocs.io/en/stable/faq.html#objective-func-additional-args).
+def objective(trial):
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:41]: Please add one more blank line to follow PEP8.
```suggestion


# FYI: Objective functions can take additional arguments
# (https://optuna.readthedocs.io/en/stable/faq.html#objective-func-additional-args).
def objective(trial):
```

File: examples/xgboost_simple.py (hunk 14 above)
Comment (toshihikoyanase) [reply examples/xgboost_simple.py:27]: Please address this comment to pass the `black` checker.

File: examples/xgboost_cv.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+(1) Execute this code directly.
+    $ python xgboost_cv.py
+
+
+(2) Execute through CLI.
+    $ STUDY_NAME=`optuna create-study --direction maximize --storage sqlite:///example.db`
+    $ optuna study optimize xgboost_cv.py objective --n-trials=100 --study $STUDY_NAME \
+      --storage sqlite:///example.db
+
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:18]: #1566 removed this instruction from the other examples. Please remove it from this file too.
```suggestion
```

Comment (sskarkhanis) [reply examples/xgboost_cv.py:18]: Fixed it as per https://github.com/optuna/optuna/issues/1394

**Comment by sskarkhanis** [comment]
I carried out all the steps below,
# Set upstream
$ git checkout master
$ git remote add upstream git://github.com/optuna/optuna.git
$ git fetch upstream
$ git merge upstream/master
$ git merge --no-ff master


before making the changes you suggested and pushed...and now it's failing on more tests...
not sure what's going wrong, perhaps I can start with a new fork and push it

**Comment by toshihikoyanase** [comment]
@sskarkhanis Thank you for your update!
Hmm, I couldn't see the merge commit of `master` branch in the commit log. I guess you might merge the master to a different feature branch.

I'm sorry but the instruction was not complete. It requires an additional step, i.e., `git checkout example_xgbCV` before `git merge --no-ff master`. Could you check it?

```console
$ git checkout master
$ git remote add upstream git://github.com/optuna/optuna.git  # This line will not be required because upstream has been set.
$ git fetch upstream
$ git merge upstream/master
$ git checkout example_xgbCV  # You'll need to check out your feature branch.
$ git merge --no-ff master
```

File: examples/xgboost_cv.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+"""
+
+
+import os
+import shutil
+import sklearn.datasets
+import sklearn.metrics
+import xgboost as xgb
+import optuna
```
Comment (toshihikoyanase) [inline examples/xgboost_cv.py:22]: Please add a blank line before optuna and the third party libraries to pass the `black` checker. In addition, please check https://github.com/optuna/optuna/pull/1202#discussion_r464736130 on `examples/xgboost_simple.py`.
```suggestion
import xgboost as xgb

import optuna
```

**Comment by sskarkhanis** [comment]
I carried out all the steps below,
# Set upstream
$ git checkout master
$ git remote add upstream git://github.com/optuna/optuna.git
$ git fetch upstream
$ git merge upstream/master
$ git merge --no-ff master


before making the changes you suggested and pushed...and now it's failing on more tests...
not sure what's going wrong, perhaps I can start with a new fork and push it

**Comment by toshihikoyanase** [comment]
@sskarkhanis Thank you for merging the master branch to your feature branch. Most of the CI jobs succeeded. The [`ci/circleci: checks`](https://circleci.com/gh/optuna/optuna/52141?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link), however, still failed due to the removal of the necessary blank line.

Please refer to [this comment](https://github.com/optuna/optuna/pull/1202#discussion_r467579212) to fix the issue.

File: examples/xgboost_cv.py (hunk 18 above)
Comment (sskarkhanis) [reply examples/xgboost_cv.py]: made the changes as you suggested but still get ci failing.

Comment (toshihikoyanase) [reply examples/xgboost_cv.py]: Hmm, I cannot see the blank line in your commit. My suggestion is out-dated, so could you add a blank line manually?
